src/test/app/AMM_test.cpp

```diff
@@ -1128,6 +1128,62 @@ struct AMM_test : public jtx::AMMTest
                 expectLedgerEntryRoot(env, carol, XRPAmount{28'999'999'990}));
         });
 
+        // equal asset deposit: unit test to exercise the rounding-down of
+        // LPTokens in the AMMHelpers.cpp: adjustLPTokens calculations
+        // The LPTokens need to have 16 significant digits and a fractional part
+        for (const Number deltaLPTokens :
+             {Number{UINT64_C(100000'0000000009), -10},
+              Number{UINT64_C(100000'0000000001), -10}})
+        {
+            testAMM([&](AMM& ammAlice, Env& env) {
+                // initial LPToken balance
+                IOUAmount const initLPToken = ammAlice.getLPTokensBalance();
+                const IOUAmount newLPTokens{
+                    deltaLPTokens.mantissa(), deltaLPTokens.exponent()};
+
+                // carol performs a two-asset deposit
+                ammAlice.deposit(
+                    DepositArg{.account = carol, .tokens = newLPTokens});
+
+                IOUAmount const finalLPToken = ammAlice.getLPTokensBalance();
+
+                // Change in behavior due to rounding down of LPTokens:
+                // there is a decrease in the observed return of LPTokens --
+                // Inputs Number{UINT64_C(100000'0000000001), -10} and
+                // Number{UINT64_C(100000'0000000009), -10} are both rounded
+                // down to 1e5
+                BEAST_EXPECT((finalLPToken - initLPToken == IOUAmount{1, 5}));
+                BEAST_EXPECT(finalLPToken - initLPToken < deltaLPTokens);
+
+                // fraction of newLPTokens/(existing LPToken balance). The
+                // existing LPToken balance is 1e7
+                const Number fr = deltaLPTokens / 1e7;
+
+                // The below equations are based on Equation 1, 2 from XLS-30d
+                // specification, Section: 2.3.1.2
+                const Number deltaXRP = fr * 1e10;
+                const Number deltaUSD = fr * 1e4;
+
+                const STAmount depositUSD =
+                    STAmount{USD, deltaUSD.mantissa(), deltaUSD.exponent()};
+
+                const STAmount depositXRP =
+                    STAmount{XRP, deltaXRP.mantissa(), deltaXRP.exponent()};
+
+                // initial LPTokens (1e7) + newLPTokens
+                BEAST_EXPECT(ammAlice.expectBalances(
+                    XRP(10'000) + depositXRP,
+                    USD(10'000) + depositUSD,
+                    IOUAmount{1, 7} + newLPTokens));
```

Comment:
I don't understand how lines 1156 and 1177 can be true at the same. I'm guessing it's due to the loss of precision in the last few digits.

`newLPTokens` == `deltaLPTokens` (numerically)
`initLPTokens` + `newLPTokens` = `finalLPTokens`, but `finalLPToken - initLPToken < deltaLPTokens`